src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Url/Base64UrlEncoder.cs

```diff
@@ -0,0 +1,500 @@
+        /// <param name="destination">The output span which contains the result of the operation, i.e. the UTF-8 encoded chars in Base64Url.</param>
+        /// <returns>The number of bytes written into the destination span. This can be used to slice the output for subsequent calls, if necessary.</returns>
+        /// <exception cref="ArgumentException">Thrown when the encoded output cannot fit in the <paramref name="destination"/> provided.</exception>
+        /// <remarks>The output will not be padded even if the input is not a multiple of 3.</remarks>
+        public static int EncodeToChars(ReadOnlySpan<byte> source, Span<char> destination)
+        {
+            OperationStatus status = EncodeToChars(source, destination, out _, out int charsWritten);
+
+            if (OperationStatus.Done == status)
+            {
+                return charsWritten;
+            }
+
+            if (OperationStatus.DestinationTooSmall == status)
+            {
+                throw new ArgumentException(SR.Argument_DestinationTooShort, nameof(destination));
+            }
+
+            Debug.Fail("Unreachable code");
+            return 0;
+        }
+
+        /// <summary>
+        /// Encode the span of binary data into UTF-8 encoded chars represented as Base64Url.
+        /// </summary>
+        /// <param name="source">The input span which contains binary data that needs to be encoded.</param>
+        /// <returns>A char array which contains the result of the operation, i.e. the UTF-8 encoded chars in Base64Url.</returns>
+        /// <remarks>The output will not be padded even if the input is not a multiple of 3.</remarks>
+        public static char[] EncodeToChars(ReadOnlySpan<byte> source)
+        {
+            char[] destination = new char[GetEncodedLength(source.Length)];
+            EncodeToChars(source, destination, out _, out int charsWritten);
+            Debug.Assert(destination.Length == charsWritten);
+
+            return destination;
+        }
+
+        /// <summary>
+        /// Encode the span of binary data into UTF-8 encoded chars represented as Base64Url.
+        /// </summary>
+        /// <param name="source">The input span which contains binary data that needs to be encoded.</param>
+        /// <returns>A string which contains the result of the operation, i.e. the UTF-8 encoded chars in Base64Url.</returns>
+        /// <remarks>The output will not be padded even if the input is not a multiple of 3.</remarks>
+        public static string EncodeToString(ReadOnlySpan<byte> source)
+        {
+            char[] destination = new char[GetEncodedLength(source.Length)];
+            EncodeToChars(source, destination, out _, out int charsWritten);
+            Debug.Assert(destination.Length == charsWritten, $"The source length: {source.Length}, bytes written: {charsWritten}");
+
+            return new string(destination);
+        }
```

Comment:
```suggestion
        public static unsafe string EncodeToString(ReadOnlySpan<byte> source)
        {
            int encodedLength = GetEncodedLength(source.Length);

#pragma warning disable CS8500 // This takes the address of, gets the size of, or declares a pointer to a managed type
            return string.Create(encodedLength, (IntPtr)(&source), static (buffer, spanPtr) =>
            {
                ReadOnlySpan<byte> source = *(ReadOnlySpan<byte>*)spanPtr;
                EncodeToChars(source, buffer, out _, out int charsWritten);
                Debug.Assert(buffer.Length == charsWritten, $"The source length: {source.Length}, bytes written: {charsWritten}");
            });
#pragma warning restore CS8500 // This takes the address of, gets the size of, or declares a pointer to a managed type
        }
```
to avoid the copy of `destination` in `new string(destination)` -- write instead to the internal string buffer directly.